### PR TITLE
Add libpng to fuzzer.

### PR DESCRIPTION
**Pull request by bungeman**
Allow fuzzing of the pngshim.c code and handling of png bitmaps.

I have locally built and run the oss-fuzz images and fuzzed for a bit. The pngshim.c file now shows up in the coverage reports. The added submodules are those used by oss-fuzz/project/zlib and oss-fuzz/project/libpng. This does not add any new fuzz targets or add anything to the base corpus, which can be done later.

### CODE REVIEW

**Comment by bungeman** [comment]
This actually isn't quite complete, turns out the patch to ftoption.h needs to be updated as well.

**Comment by bungeman** [comment]
Turns out while the ftoptions.h patch should be updated, I'm not seeing any of the fuzz targets exercising the png sbit code even when handed a valid sbix font with embedded png data. Will need to update the fuzz target as well.

**Comment by bungeman** [comment]
Also wanted to point out that the `TRAVIS_CI_ROW="oss-fuzz-build"` job is not useful, as it always uses whatever is on the master branch of freetype/freetype2-testing , it is not testing this change at all, which is unfortunate.

**Comment by bungeman** [comment]
At this point this pull request adds a small sbix and a small CBDT font to the corpus and I can verify that the `driver truetype-renderer` will go through `Load_SBit_Png` (ran under gdb and set a breakpoint). I have a local oss-fuzz hacked up to use the pull request branch as its source and created the docker image and ran the fuzzer. However, while `pngshim.h` shows up in the coverage report no coverage of its content is reported. It's possible that something is just incorrect with how I set up my local oss-fuzz, so it may be worth landing this anyway since it  extends the possible fuzzing.

**Comment by bungeman** [comment]
So I cut down the size of the already small sbix and cbdt fonts to ~2k and now I'm sure the `driver truetype-renderer` is actually running the png code at least sometimes because I just got it working and got a leak report. Will investigate that.

**Comment by bungeman** [comment]
I believe this is now ready to land.

I took a look into the `TRAVIS_CI_ROW="oss-fuzz-build"` and while it is clear what needs to be done, it isn't clear how or if it is possible to plumb the needed information through the layers. Ideally there would be some way for the oss-fuzz/project/freetype2/Dockerfile to take an argument so that when freetype2-testing is cloned the merge commit under consideration can be fetched and checked out. Either that or a later step in the process could do this.

**Comment by cherusker** [comment]
oh lovely PR @bungeman, thanks a lot! 😃 I'll come back to you about everything asap! 😊

**Comment by bungeman** [comment]
The leak found was reported as https://savannah.nongnu.org/bugs/?59322 and fixed with https://git.savannah.gnu.org/cgit/freetype/freetype2.git/commit/?id=007c109b4594c5e63948bd08b4d5011ad76ffb10 .

**Comment by cherusker** [comment]
just a quick headsup @bungeman - I'll take a look at your PR tomorrow + am very hopeful we can merge it in _as is_ 😊 👍 ... just quickly want to take a look at building the two libraries myself; sorry that this is taking so long!

File: fuzzing/scripts/build/zlib.sh
Comment (cherusker) [inline fuzzing/scripts/build/zlib.sh:32]: Let's add `--static` here since OSS-Fuzz will only ever use the static builds. That way we can avoid mixing up the two artefacts down the road.

File: fuzzing/scripts/build/libpng.sh
Comment (cherusker) [inline fuzzing/scripts/build/libpng.sh:38]: oh noo haha I believe there is not other way to get around this? 🙈 ... out of curiosity -- is there a specific reason why we wouldn't want to use `sed -i`?

File: fuzzing/scripts/build/libpng.sh
Comment (cherusker) [inline fuzzing/scripts/build/libpng.sh:45]: A few quick thoughts here:

- Calling `configure` and creating the dirs are two quite distinct things to do (esp. as the `configure` call spans across multiple lines); if you don't mind I'd appreciated an empty line between `mkdir ...` + `CPPFLAGS ...` 😊 

- I generally try to use the `${}` form for accessing variable contents wherever possible (esp. b/c I'm no `bash` expert by any means and general concensus seems to be that that would be the "safest" way forward) -- happy to hear your thoughts on this 😊 

- let's please use the `--enable-shared` flag to disable the shared build here as well

**Review by cherusker — COMMENTED** [review]
Sorry that it took me so long; there are just a lot of variables to this PR and I want to make sure I'm giving it all the attention it deserves! 😊

Also, thanks again for putting together this PR + digging through the repo to find out how things are structured and organised! 💛 

I just put down a few quick thoughts and have one question about the `libpng` commit -- is there are reason why we wouldn't want to use https://github.com/glennrp/libpng/releases/tag/v1.6.37? (I have a suspicion based on that commit title but wanted to double check.)

**Review by bungeman — COMMENTED** [review]
As far as png version I think it's just at most recent, since this would need to keep up with any fixes in libpng found oss-fuzz. It would be nice if there were a clean way in oss-fuzz to state that a fuzz project depends on other fuzz projects and would like to use the most recent good build of that target (like a package manager full of packages with various sanitizer builds).

File: fuzzing/scripts/build/libpng.sh
Comment (cherusker) [reply fuzzing/scripts/build/libpng.sh:40]: > maybe a comment about where this comes from

this sounds great I think; if you don't mind adding that I'd appreciated it 😊

**Review by cherusker — APPROVED** [review]
Just tested it and it looks great; thanks a lot for this addition! 😊 and sorry again that it took several weeks to complete!

Let me know if you'd like to add a comment to the `sed` usage in the build script (I'm fine either way); and let's merge it then 🎉

**Comment by cherusker** [comment]
> It would be nice if there were a clean way in oss-fuzz to state that a fuzz project depends on other fuzz projects and would like to use the most recent good build of that target (like a package manager full of packages with various sanitizer builds).

I couldn't agree more! 😊

File: fuzzing/scripts/build/libpng.sh
Comment (bungeman) [reply fuzzing/scripts/build/libpng.sh:40]: Added comment already up at line 32 with a link to the build file this was sourced from, so that if anything ever stops working here we know where to look.

File: fuzzing/scripts/build/libpng.sh
Comment (bungeman) [inline fuzzing/scripts/build/libpng.sh:33]: Added a comment here so that if anything ever stops working we know where to look. (It seems my comment in the conversation on the old pull request commit still shows up under the old pull request commit, even if I made that comment while responding to the conversation on the new pull request commit. So mentioning this here so that it is attached to the newer commit.)

**Review by bungeman — COMMENTED** [review]
I am used to how changes and conversations work in Gerrit. I am confused by the presentation here, so attempting to comment on the right thing...

**Comment by bungeman** [comment]
I think this should be good to go now, but the status box here tells me that while everything has been approved I don't have write access so cannot land this myself.

**Comment by cherusker** [comment]
> I think this should be good to go now, but the status box here tells me that while everything has been approved I don't have write access so cannot land this myself.

I'm super sorry, I thought that was fixed! 😞 I'll look into this; I think there should be a mode where contributors are able to merge code that passes all checks -- but not used to setting up open source GitHub repos myself (and started from the viewpoint of "most restricted" with the idea of making it "more open" as required); very happy for hints if you have ideas! 😊

**Pull request merged**
